TSM1 Engine - panic: runtime error: index out of range - Values.MinTime

Hello,

After starting from scratch InfluxDB (87d99fb) after around 24h's I have got the following panic.

```
2015/10/13 23:13:51 InfluxDB starting, version 0.9, branch master, commit 87d99fb, built 1444754119
2015/10/13 23:13:51 Go version go1.5.1, GOMAXPROCS set to 8
2015/10/13 23:13:51 Using configuration at: /opt/influxdb/influxd.conf
[metastore] 2015/10/13 23:13:51 Using data dir: /opt/graphite/storage/influxdb/meta
...
[continuous_querier] 2015/10/14 09:22:16 wrote 56 point(s) to ESXi.raw
[tsm1wal] 2015/10/14 09:22:45 memory flush of /opt/graphite/storage/influxdb/data/ESXi/raw/4 with 8679 keys and 359602 total values of 5753632 bytes
[tsm1] 2015/10/14 09:22:46 rewriting index file /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001371.tsm1 with /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001379.tsm1
[tsm1] 2015/10/14 09:22:46 rewriting index file /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001377.tsm1 with /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001380.tsm1
panic: runtime error: index out of range

goroutine 62062 [running]:
github.com/influxdb/influxdb/tsdb/engine/tsm1.Values.MinTime(0xc863a91e80, 0x0, 0x0, 0x140cfb7e37d14800)
        /opt/gcollectors/src/go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/encoding.go:65 +0x75
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*Engine).DecodeAndCombine(0xc846484a20, 0xc863a91e80, 0x0, 0x0, 0x9d5eec64e, 0x2d, 0x89f69bd, 0xc838f60000, 0x0, 0x4e20, ...)
        /opt/gcollectors/src/go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/tsm1.go:1668 +0x6bd
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*Engine).rewriteFile(0xc846484a20, 0xc8621de9c0, 0xc8622c8060, 0x0, 0x0)
        /opt/gcollectors/src/go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/tsm1.go:1111 +0x208f
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*Engine).Write.func3(0xc846484a20, 0xc863cadf40, 0xc850051480, 0xc863cadf50, 0xc8621de9c0, 0xc8622c8060)
        /opt/gcollectors/src/go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/tsm1.go:415 +0x3f
created by github.com/influxdb/influxdb/tsdb/engine/tsm1.(*Engine).Write
        /opt/gcollectors/src/go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/tsm1.go:421 +0x95c
goroutine 1 [chan receive, 607 minutes]:
main.(*Main).Run(0xc8201f7f10, 0xc82008a010, 0x1, 0x1, 0x0, 0x0)
        /opt/gcollectors/src/go/src/github.com/influxdb/influxdb/cmd/influxd/main.go:96 +0xa39
main.main()
        /opt/gcollectors/src/go/src/github.com/influxdb/influxdb/cmd/influxd/main.go:46 +0x29b

goroutine 17 [syscall, 607 minutes, locked to thread]:
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 20 [syscall, 607 minutes]:
os/signal.loop()
        /usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
        /usr/local/go/src/os/signal/signal_unix.go:28 +0x37
goroutine 37 [select]:
github.com/hashicorp/raft.(*Raft).leaderLoop(0xc8201ca000)
        /opt/gcollectors/src/go/src/github.com/hashicorp/raft/raft.go:849 +0xcce
github.com/hashicorp/raft.(*Raft).runLeader(0xc8201ca000)
        /opt/gcollectors/src/go/src/github.com/hashicorp/raft/raft.go:821 +0x936
github.com/hashicorp/raft.(*Raft).run(0xc8201ca000)
        /opt/gcollectors/src/go/src/github.com/hashicorp/raft/raft.go:602 +0xbb
github.com/hashicorp/raft.(*Raft).(github.com/hashicorp/raft.run)-fm()
        /opt/gcollectors/src/go/src/github.com/hashicorp/raft/raft.go:252 +0x20
github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc8201ca000, 0xc820122be0)
        /opt/gcollectors/src/go/src/github.com/hashicorp/raft/state.go:152 +0x44
created by github.com/hashicorp/raft.(*raftState).goFunc
```

```
# ls - /opt/graphite/storage/influxdb/data/ESXi/raw/4/00013*
-rw-r--r--  1 root  influxd   3.5M Oct 14 08:55 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001300.tsm1
-rw-r--r--  1 root  influxd    11M Oct 14 08:55 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001301.tsm1
-rw-r--r--  1 root  influxd   3.9M Oct 14 08:55 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001302.tsm1
-rw-r--r--  1 root  influxd    11M Oct 14 08:55 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001303.tsm1
-rw-r--r--  1 root  influxd    11M Oct 14 08:58 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001304.tsm1
-rw-r--r--  1 root  influxd   120M Oct 14 08:59 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001305.tsm1
-rw-r--r--  1 root  influxd   9.0M Oct 14 09:01 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001321.tsm1
-rw-r--r--  1 root  influxd    11M Oct 14 09:01 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001322.tsm1
-rw-r--r--  1 root  influxd   688K Oct 14 09:01 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001323.tsm1
-rw-r--r--  1 root  influxd   1.1M Oct 14 09:02 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001329.tsm1
-rw-r--r--  1 root  influxd   9.8M Oct 14 09:02 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001330.tsm1
-rw-r--r--  1 root  influxd    11M Oct 14 09:02 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001331.tsm1
-rw-r--r--  1 root  influxd   120M Oct 14 09:05 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001332.tsm1
-rw-r--r--  1 root  influxd    11M Oct 14 09:08 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001337.tsm1
-rw-r--r--  1 root  influxd   2.8M Oct 14 09:08 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001340.tsm1
-rw-r--r--  1 root  influxd    10M Oct 14 09:08 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001341.tsm1
-rw-r--r--  1 root  influxd   3.1M Oct 14 09:10 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001342.tsm1
-rw-r--r--  1 root  influxd    10M Oct 14 09:10 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001343.tsm1
-rw-r--r--  1 root  influxd    10M Oct 14 09:12 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001344.tsm1
-rw-r--r--  1 root  influxd   921K Oct 14 09:15 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001368.tsm1
-rw-r--r--  1 root  influxd   9.8M Oct 14 09:15 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001369.tsm1
-rw-r--r--  1 root  influxd    10M Oct 14 09:15 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001370.tsm1
-rw-r--r--  1 root  influxd   1.1M Oct 14 09:15 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001371.tsm1
-rw-r--r--  1 root  influxd    10M Oct 14 09:16 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001374.tsm1
-rw-r--r--  1 root  influxd   9.9M Oct 14 09:16 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001375.tsm1
-rw-r--r--  1 root  influxd   9.9M Oct 14 09:18 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001376.tsm1
-rw-r--r--  1 root  influxd   138M Oct 14 09:21 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001377.tsm1
-rw-r--r--  1 root  influxd   9.9M Oct 14 09:20 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001378.tsm1
-rw-r--r--  1 root  influxd   698K Oct 14 09:22 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001379.tsm1
-rw-r--r--  1 root  influxd     0B Oct 14 09:22 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001379.tsm1.check
-rw-r--r--  1 root  influxd    50K Oct 14 09:22 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001380.tsm1
-rw-r--r--  1 root  influxd     0B Oct 14 09:22 /opt/graphite/storage/influxdb/data/ESXi/raw/4/0001380.tsm1.check
```

If you need more details don't hesitate in asking.

Regards,
Seabstian

--- a/cmd/influx_inspect/tsm.go
+++ b/cmd/influx_inspect/tsm.go
@@ -165,11 +165,11 @@ func readIds(path string) (map[string]uint64, error) {
 	}
 	return ids, err
 }
-func readIndex(f *os.File) *tsmIndex {
+func readIndex(f *os.File) (*tsmIndex, error) {
 	// Get the file size
 	stat, err := f.Stat()
 	if err != nil {
-		panic(err.Error())
+		return nil, err
 	}
 
 	// Seek to the series count
@@ -177,8 +177,7 @@ func readIndex(f *os.File) *tsmIndex {
 	b := make([]byte, 8)
 	_, err = f.Read(b[:4])
 	if err != nil {
-		fmt.Printf("error: %v\n", err.Error())
-		os.Exit(1)
+		return nil, err
 	}
 
 	seriesCount := binary.BigEndian.Uint32(b)
@@ -206,6 +205,10 @@ func readIndex(f *os.File) *tsmIndex {
 		series:  count,
 	}
 
+	if indexStart < 0 {
+		return nil, fmt.Errorf("index corrupt: offset=%d", indexStart)
+	}
+
 	// Read the index entries
 	for i := 0; i < count; i++ {
 		f.Read(b)
@@ -215,7 +218,7 @@ func readIndex(f *os.File) *tsmIndex {
 		index.blocks = append(index.blocks, &block{id: id, offset: int64(pos)})
 	}
 
-	return index
+	return index, nil
 }
 
 func cmdDumpTsm1(opts *tsdmDumpOpts) {
@@ -254,7 +257,19 @@ func cmdDumpTsm1(opts *tsdmDumpOpts) {
 		invIds[v] = k
 	}
 
-	index := readIndex(f)
+	index, err := readIndex(f)
+	if err != nil {
+		println("Failed to readIndex:", err.Error())
+
+		// Create a stubbed out index so we can still try and read the block data directly
+		// w/o panicing ourselves.
+		index = &tsmIndex{
+			minTime: time.Unix(0, 0),
+			maxTime: time.Unix(0, 0),
+			offset:  stat.Size(),
+		}
+	}
+
 	blockStats := &blockStats{}
 
 	println("Summary:")

--- a/tsdb/engine/tsm1/tsm1.go
+++ b/tsdb/engine/tsm1/tsm1.go
@@ -1649,6 +1649,11 @@ func (e *Engine) readSeries() (map[string]*tsdb.Series, error) {
 // has future encoded blocks so that this method can know how much of its values can be
 // combined and output in the resulting encoded block.
 func (e *Engine) DecodeAndCombine(newValues Values, block, buf []byte, nextTime int64, hasFutureBlock bool) (Values, []byte, error) {
+	// No new values passed in, so nothing to combine.  Just return the existing block.
+	if len(newValues) == 0 {
+		return newValues, block, nil
+	}
+
 	values, err := DecodeBlock(block)
 	if err != nil {
 		panic(fmt.Sprintf("failure decoding block: %v", err))

--- a/tsdb/engine/tsm1/tsm1_test.go
+++ b/tsdb/engine/tsm1/tsm1_test.go
@@ -1013,7 +1013,81 @@ func TestEngine_WriteIntoCompactedFile(t *testing.T) {
 	}
 
 	if count := e.DataFileCount(); count != 1 {
-		t.Fatalf("execpted 1 data file but got %d", count)
+		t.Fatalf("expected 1 data file but got %d", count)
+	}
+
+	tx, _ := e.Begin(false)
+	defer tx.Rollback()
+	c := tx.Cursor("cpu,host=A", fields, nil, true)
+	k, _ := c.SeekTo(0)
+	if k != 1000000000 {
+		t.Fatalf("wrong time: %d", k)
+	}
+	k, _ = c.Next()
+	if k != 2000000000 {
+		t.Fatalf("wrong time: %d", k)
+	}
+	k, _ = c.Next()
+	if k != 2500000000 {
+		t.Fatalf("wrong time: %d", k)
+	}
+	k, _ = c.Next()
+	if k != 3000000000 {
+		t.Fatalf("wrong time: %d", k)
+	}
+	k, _ = c.Next()
+	if k != 4000000000 {
+		t.Fatalf("wrong time: %d", k)
+	}
+}
+
+func TestEngine_WriteIntoCompactedFile_MaxPointsPerBlockZero(t *testing.T) {
+	e := OpenDefaultEngine()
+	defer e.Close()
+
+	fields := []string{"value"}
+
+	e.MaxPointsPerBlock = 4
+	e.RotateFileSize = 10
+
+	p1 := parsePoint("cpu,host=A value=1.1 1000000000")
+	p2 := parsePoint("cpu,host=A value=1.2 2000000000")
+	p3 := parsePoint("cpu,host=A value=1.3 3000000000")
+	p4 := parsePoint("cpu,host=A value=1.5 4000000000")
+	p5 := parsePoint("cpu,host=A value=1.6 2500000000")
+	p6 := parsePoint("cpu,host=A value=1.7 5000000000")
+	p7 := parsePoint("cpu,host=A value=1.8 6000000000")
+	p8 := parsePoint("cpu,host=A value=1.9 7000000000")
+
+	if err := e.WritePoints([]models.Point{p1, p2}, nil, nil); err != nil {
+		t.Fatalf("failed to write points: %s", err.Error())
+	}
+	if err := e.WritePoints([]models.Point{p3}, nil, nil); err != nil {
+		t.Fatalf("failed to write points: %s", err.Error())
+	}
+
+	if err := e.Compact(true); err != nil {
+		t.Fatalf("error compacting: %s", err.Error())
+	}
+
+	if err := e.WritePoints([]models.Point{p4}, nil, nil); err != nil {
+		t.Fatalf("failed to write points: %s", err.Error())
+	}
+
+	if err := e.WritePoints([]models.Point{p6, p7, p8}, nil, nil); err != nil {
+		t.Fatalf("failed to write points: %s", err.Error())
+	}
+
+	if err := e.Compact(true); err != nil {
+		t.Fatalf("error compacting: %s", err.Error())
+	}
+
+	if err := e.WritePoints([]models.Point{p5}, nil, nil); err != nil {
+		t.Fatalf("failed to write points: %s", err.Error())
+	}
+
+	if count := e.DataFileCount(); count != 1 {
+		t.Fatalf("expected 1 data file but got %d", count)
 	}
 
 	tx, _ := e.Begin(false)
@@ -1351,6 +1425,32 @@ func TestEngine_RewriteFileAndCompact(t *testing.T) {
 			t.Fatalf("wrong time %d", k)
 		}
 	}()
+}
+
+func TestEngine_DecodeAndCombine_NoNewValues(t *testing.T) {
+	var newValues tsm1.Values
+	e := OpenDefaultEngine()
+	defer e.Engine.Close()
+
+	values := make(tsm1.Values, 1)
+	values[0] = tsm1.NewValue(time.Unix(0, 0), float64(1))
+
+	block, err := values.Encode(nil)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	remaining, encoded, err := e.DecodeAndCombine(newValues, block, nil, time.Unix(1, 0).UnixNano(), false)
+	if len(remaining) != 0 {
+		t.Fatalf("unexpected remaining values: exp %v, got %v", 0, len(remaining))
+	}
+
+	if len(encoded) != len(block) {
+		t.Fatalf("unexpected encoded block length: exp %v, got %v", len(block), len(encoded))
+	}
+
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
 }
 
 // Engine represents a test wrapper for tsm1.Engine.

